slave-base/contrib/bin/run-jnlp-client

```diff
@@ -65,57 +63,88 @@ if [[ $# -gt 1 ]]; then
+        alternatives --set java $JVMPath64bit
+      else
+        echo "max heap in MB is ${CONTAINER_HEAP_MAX} and 64 bit was not explicitly set so using 32 bit Java"
+        alternatives --set java $JVMPath32bit
+        export MALLOC_ARENA_MAX=${MALLOC_ARENA_MAX:-1}
+      fi
     fi
-    CONTAINER_INITIAL_HEAP=$(echo "${CONTAINER_HEAP_MAX} ${CONTAINER_INITIAL_PERCENT}" | awk '{ printf "%d", $1 * $2 }')
-    JAVA_INITIAL_HEAP_PARAM="-Xms${CONTAINER_INITIAL_HEAP}m"
-  fi
 
-  if [ -z $JAVA_GC_OPTS ]; then
-      # We no longer set MaxMetaspaceSize because the JVM should expand metaspace until it reaches the container limit.
-      # See http://hg.openjdk.java.net/jdk8u/jdk8u/hotspot/file/4dd24f4ca140/src/share/vm/memory/metaspace.cpp#l1470
-      JAVA_GC_OPTS="-XX:+UseParallelGC -XX:MinHeapFreeRatio=5 -XX:MaxHeapFreeRatio=10 -XX:GCTimeRatio=4 -XX:AdaptiveSizePolicyWeight=90"
+  else
+    # no k8s/docker memory limit set
+    if [[ -z $OPENSHIFT_JENKINS_JVM_ARCH ]]; then
+      echo "no cgroup memory limit found, using 64 bit Java"
+      alternatives --set java $JVMPath64bit
+    fi
   fi
 
+  if [[ -z "$JAVA_GC_OPTS" ]]; then
+    # See https://developers.redhat.com/blog/2014/07/22/dude-wheres-my-paas-memory-tuning-javas-footprint-in-openshift-part-2/ .
+    # The values are aggressively set with the intention of relaxing GC CPU time
+    # restrictions to enable it to free as much as possible, as well as
+    # encouraging the GC to free unused heap memory back to the OS.
+    JAVA_GC_OPTS="-XX:+UseParallelGC -XX:MinHeapFreeRatio=5 -XX:MaxHeapFreeRatio=10 -XX:GCTimeRatio=4 -XX:AdaptiveSizePolicyWeight=90"
+  fi
 
-  if [ ! -z "${USE_JAVA_DIAGNOSTICS}" ]; then
-      JAVA_DIAGNOSTICS="-XX:NativeMemoryTracking=summary -XX:+PrintGC -XX:+PrintGCDateStamps -XX:+PrintGCTimeStamps -XX:+UnlockDiagnosticVMOptions"
+  # USE_JAVA_DIAGNOSTICS and JAVA_DIAGNOSTICS are legacy and may be removed in
+  # a future version of this script.
+  if [[ "${USE_JAVA_DIAGNOSTICS}" ]]; then
+    JAVA_DIAGNOSTICS="-XX:NativeMemoryTracking=summary -XX:+PrintGC -XX:+PrintGCDateStamps -XX:+PrintGCTimeStamps -XX:+UnlockDiagnosticVMOptions"
   fi
 
-  if [ ! -z "${CONTAINER_CORE_LIMIT}" ]; then
-      JAVA_CORE_LIMIT="-XX:ParallelGCThreads=${CONTAINER_CORE_LIMIT} -Djava.util.concurrent.ForkJoinPool.common.parallelism=${CONTAINER_CORE_LIMIT} -XX:CICompilerCount=2"
+  if [[ "${CONTAINER_CORE_LIMIT}" ]]; then
+    JAVA_CORE_LIMIT="-XX:ParallelGCThreads=${CONTAINER_CORE_LIMIT} -Djava.util.concurrent.ForkJoinPool.common.parallelism=${CONTAINER_CORE_LIMIT} -XX:CICompilerCount=2"
   fi
 
-  if [ -z "${JAVA_OPTS}" ]; then
-      JAVA_OPTS="-XX:+UnlockExperimentalVMOptions -XX:+UseCGroupMemoryLimitForHeap -Dsun.zip.disableMemoryMapping=true"
+  if [[ -z "${CONTAINER_JAVA_OPTIONS}" ]]; then
+    CONTAINER_JAVA_OPTIONS="$JAVA_GC_OPTS $JAVA_INITIAL_HEAP_PARAM $JAVA_MAX_HEAP_PARAM $JAVA_CORE_LIMIT $JAVA_DIAGNOSTICS"
```

Comment:
JNLP_JAVA_OPTIONS.  Same elsewhere as applicable.

Reply:
Will do - am assuming to keep the already established CONTAINER_CORE_LIMIT / CONTAINER_INITIAL_PERCENT etc. as they are.